gtsfm/utils/images.py

```diff
@@ -118,6 +118,45 @@ def resize_image(image: Image, new_height: int, new_width: int) -> Image:
     return Image(resized_value_array)
 
 
+def get_downsample_factor_per_axis(img: Image, max_resolution: int) -> Tuple[float, float, int, int]:
+    """
+    Resizing an image by a specific downsample ratio may not be possible due to lack of a clean
+    divisor. However, we can still determine the exact downsampling ratio.
+
+    Args:
+        img: image to be downsampled
+        max_resolution: integer representing maximum length of image's short side
+            e.g. for 1080p (1920 x 1080), max_resolution would be 1080
+
+    Returns:
+        downsample_u: downsampling factor for u coordinate
+        downsample_v: downsampling factor for v coordinate
+        new_h: new height that will preserve aspect ratio as closely as possible, while
+            respecting max_resolution constraint
+        new_w: new width
+    """
+    h, w = img.height, img.width
+    shorter_size = min(h, w)
+    if shorter_size == h:
+        new_h = max_resolution
+        # compute scaling that will be applied to original image
+        scale = new_h / float(h)
+        new_w = int(w * scale)
+    else:
+        new_w = max_resolution
+        scale = new_w / float(w)
+        new_h = int(h * scale)
+
+    downsampled_img = resize_image(img, new_height=new_h, new_width=new_w)
```

Comment:
I don’t get why we are doing the resizing here..

Reply:
So the main idea is that OpenCV will do some rounding on its end -- and unless we dig into its source to see its protocol, I can't predict it

```
def test_get_downsample_factor_per_axis() -> None:
    """Ensure that max resolution constraint is met, when downsampling image.
    Resize a 700x1500 image, so that the shorter image side is at most 600 px.
    """
    img = Image(np.zeros((700, 1500, 3), dtype=np.uint8))
    max_resolution = 600
    downsample_u, downsample_v, new_h, new_w = image_utils.get_downsample_factor_per_axis(img, max_resolution)

    # 6/7 will not give a clean integer division
    assert new_h == 1.167
    assert new_w == 1.166
    assert downsample_u == 600
    assert downsample_v == 1285
```

Reply:
I understand this finally after our offline discussion. The function returns the scaling factors and the new height and width. It also calls the OpenCV resize function to check that this new height and width is actually "feasible". 

Firstly, I dont think the function documentation states this. But more importantly, I dont think it is good design to have a computationally expensive check inside an otherwise computationally cheap util function. Maybe the name and outputs of the function can be changed? 

Curious to know what @ayushbaid thinks to see if this is just my personal opinion.